backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/loops/ProgressionHandlers.kt

```diff
@@ -0,0 +1,148 @@
+        singleArgumentExtension(FqName("kotlin.ranges.downTo"), progressionElementClasses)
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType): ProgressionInfo? =
+            ProgressionInfo(progressionType,
+                    call.extensionReceiver!!,
+                    call.getValueArgument(0)!!,
+                    increasing = false)
+}
+
+internal class UntilHandler(val progressionElementClasses: Collection<IrClassSymbol>) : ProgressionHandler {
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.until"), progressionElementClasses)
+        parameterRestriction(0) { it.type.classifierOrNull in progressionElementClasses }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType): ProgressionInfo? =
+            ProgressionInfo(progressionType,
+                    call.extensionReceiver!!,
+                    call.getValueArgument(0)!!,
+                    closed = false)
+}
+
+internal class IndicesHandler(val context: Context) : ProgressionHandler {
+
+    private val symbols = context.ir.symbols
+
+    private val supportedArrays = symbols.primitiveArrays.values + symbols.array
+
+    override val matcher = SimpleCalleeMatcher {
+        extensionReceiverRestriction { it != null && it.type.classifierOrNull in supportedArrays }
+        fqNameRestriction { it == FqName("kotlin.collections.<get-indices>") }
+        parameterCountRestriction { it == 0 }
+    }
+
+    override fun build(call: IrCall, progressionType: ProgressionType): ProgressionInfo? {
+        val int0 = IrConstImpl.int(call.startOffset, call.endOffset, context.irBuiltIns.intType, 0)
+
+        val bound = with(context.createIrBuilder(call.symbol, call.startOffset, call.endOffset)) {
+            val clazz = call.extensionReceiver!!.type.classifierOrFail
+            val symbol = symbols.arraySize[clazz]!!
+            irCall(symbol).apply {
+                dispatchReceiver = call.extensionReceiver
+            }
+        }
+        return ProgressionInfo(progressionType, int0, bound, closed = false)
+    }
+}
+
+internal class StepHandler(context: Context, val visitor: IrElementVisitor<ProgressionInfo?, Nothing?>) : ProgressionHandler {
+    private val symbols = context.ir.symbols
+
+    val calleeMatcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.step"), symbols.progressionClasses)
+        parameterRestriction(0) { it.type.isInt() || it.type.isLong() }
+    }
+
+    override val matcher: IrCallMatcher = object : IrCallMatcher {
+        override fun invoke(call: IrCall): Boolean {
+            // Due to KT-27607 nested steps could lead to incorrect behaviour.
+            // So disable optimization of such rare cases for now.
+            if (call.extensionReceiver is IrCall && calleeMatcher(call.extensionReceiver as IrCall)) {
```

Comment:
Can we check that `nestedInfo` has default step instead?
It looks as more reliable solution to me.